lib/msf/core/handler/bind_aws_ssm.rb

```diff
@@ -0,0 +1,370 @@
+        rescue ::Exception => e
+          ilog("AwsSsmSession monitor thread raised #{e.class}: #{e}")
+        end
+      }
+    end
+
+    # Find command response on cursor and return to caller - doesn't respect length arg, yet
+    def ssm_read(length = nil, opts = {})
+      maxw = opts[:timeout] ? opts[:timeout] : 30
+      start = Time.now
+      resp = @ssmclient.list_command_invocations(command_id: @cursor, instance_id: @peer_info['InstanceId'], details: true)
+      while (resp.command_invocations.empty? or resp.command_invocations[0].status == "InProgress") and
+        (Time.now - start).to_i.abs < maxw do
+        Rex::ThreadSafe.sleep(1)
+        resp = @ssmclient.list_command_invocations(command_id: @cursor, instance_id: @peer_info['InstanceId'], details: true)
+      end
+      # SSM script invocation states are: InProgress, Success, TimedOut, Cancelled, Failed
+      if resp.command_invocations[0].status == "Success" or  resp.command_invocations[0].status == "Failed"
+        # The big limitation: SSM command outputs are only 2500 chars max, otherwise you have to write to S3 and read from there
+        output = resp.command_invocations.map {|c| c.command_plugins.map {|p| p.output}.join}.join
+        @cursor = nil
+        return output
+      else
+        @cursor = nil
+        ilog("AwsSsmSession error #{resp}")
+        raise resp
+      end
+      nil
+    end
+
+    def write(buf, opts = {})
+      resp = @ssmclient.send_command(
+        document_name: @cmd_doc,
+        instance_ids: [@peer_info['InstanceId']],
+        parameters: { commands: [buf] }
+      )
+      if resp.command.error_count == 0
+        @cursor = resp.command.command_id
+        return buf.length
+      else
+        @cursor = nil
+        ilog("AwsSsmSession error #{resp}")
+        raise resp
+      end
+    end
+
+    #
+    # Closes the stream abstraction and kills the monitor thread.
+    #
+    def close
+      @monitor_thread.kill if (@monitor_thread)
+      @monitor_thread = nil
+
+      cleanup_abstraction
+    end
+  end
+  #
+  # Returns the handler specific string representation, in this case
+  # 'bind_tcp'.
+  #
+  def self.handler_type
+    return "bind_aws_ssm"
+  end
+
+  #
+  # Returns the connection oriented general handler type, in this case bind.
+  #
+  def self.general_handler_type
+    "bind"
+  end
+
+  # A string suitable for displaying to the user
+  #
+  # @return [String]
+  def human_name
+    "bind AWS SSM"
+  end
+
+  #
+  # Initializes a bind handler and adds the options common to all bind
+  # payloads, such as local port.
+  #
+  def initialize(info = {})
+    super
+
+    register_options(
+      [
+        OptString.new('AWS_EC2_ID', [true, 'The EC2 ID of the instance ', '']),
+        OptString.new('AWS_REGION', [true, 'AWS region containing the instance', 'us-east-1']),
+        OptString.new('AWS_AK', [false, 'AWS access key', nil]),
+        OptString.new('AWS_SK', [false, 'AWS secret key', nil]),
```

Comment:
Any reason to not use consistent names for these in the `enum_ssm` module too? I see `ACCESS_KEY` / `SECRET_ACCESS_KEY` is consistent with enum_ec2 but I see no current uses of `AWS_AK` / `AWS_SK`.

Reply:
Yeah, i wrote it completely separately - should probably have a mixin for AWS stuff.